### ISSUE TITLE
Documentation for the new API introduced in 0.9.0

### ISSUE THREAD

**Issue by 1Maxnet1** [motis-project/motis#345]
I wanted to give the new API a try and adapt my python client to it. However I cannot find any information regarding the new API. Neither the announced OpenAPI specification, nor an update documentation. I tried to make it work based on the example in the blog post, but the returning error does only says malformed message. Is there any documentation available regarding the new API ?

Thanks in advance

**Comment by felixguendling** [comment]
In case you already have a working client for the "old" version, there's no need to switch to the new version. As long as we support Flatbuffers, both versions will be supported. The original version is not deprecated and there's no advantage of the new version except the fact that it enables us to produce an OpenAPI spec for it.

The OpenAPI playground is available in each running MOTIS instance via `/openapi`. Here the example for the demo server: https://europe.motis-project.de/openapi/#/

The MOTIS download archive for every platform contains the OpenAPI `yaml` files in the `/motis/web/openapi/` folder (`openapi.yaml` and `openapi-3.0.yaml`). So it's also hosted by default:

  - https://europe.motis-project.de/openapi/openapi.yml
  - https://europe.motis-project.de/openapi/openapi-3.0.yml

However, there is no need to switch since the original API is neither deprecated nor are there plans to deprecate it.

**Comment by felixguendling** [comment]
We have no experience with creating clients from this spec yet. So there might be rough edges if you want to use an OpenAPI-generated MOTIS client.

**Comment by felixguendling** [comment]
> but the returning error does only says malformed message

If you're running your own server where have access to the output MOTIS produces, the standard output log might give some more information what's wrong.

**Comment by 1Maxnet1** [comment]
> In case you already have a working client for the "old" version, there's no need to switch to the new version.

It used to work. However after updating the server, it stopped working that's why I considered the old api unsupported.

> If you're running your own server where have access to the output MOTIS produces, the standard output log might give some more information what's wrong.

I will have a look at the logs and check whether I can find the issue.

> We have no experience with creating clients from this spec yet. So there might be rough edges if you want to use an OpenAPI-generated MOTIS client.

This sounds like an interesting approach for the future. I will put it on my list, but won't look into it right now.

Thanks for all the information and answers, I will report back whether I can find my issue.

**Comment by felixguendling** [comment]
At one point, the `IntermodalRoutingRequest` gained the `router` field which might lead to your problem. The `router` field allows you to select the routing algorithm the `intermodal` module should use. If left empty (it can't be missing!), the algorithm chosen in the config parameter `intermodal.router` is used (which is the `routing` module by default at the moment).

**Comment by 1Maxnet1** [comment]
> At one point, the `IntermodalRoutingRequest` gained the `router` field which might lead to your problem. The `router` field allows you to select the routing algorithm the `intermodal` module should use. If left empty (it can't be missing!), the algorithm chosen in the config parameter `intermodal.router` is used (which is the `routing` module by default at the moment).

Thanks, This was wrong, thanks! (Very impressive how you can guess the debugging for my erros, without me properly reporting them ^^) However there is another issue, as it cannot find the stations anymore. I did change the config.ini to include a tag name for the loaded schedule, as motis output it was required: `paths=schedule-vvs:/data/vvs-motis/gtfs
`. However it does now fail with:
```
[erro][2023-06-21T10:37:47Z+0000][location.cc:32] nigiri: could not find de:08118:7404:1:3, de:08118:7404:1:3, 255, tags: [
  vvs_
]
[erro][2023-06-21T10:37:47Z+0000][dispatcher.cc:127] error executing /lookup/station_location: ankerl::unordered_dense::map::at(): key not found
```

If you have another idea what could be wrong let me know.

**Comment by felixguendling** [comment]
Made the same mistake myself a couple of times ;-)

When using tags to distinguish different timetables (which is required with `nigiri`, even if you have only one dataset), you need to prefix your station IDs with those tags. The reason is that different timetable datasets sometimes use the same station ID, so a station ID would not be unique anymore without the prepended tag. In your case, the station `de:08118:7404:1:3` from your dataset would be `vvs_de:08118:7404:1:3` in MOTIS (depending on the tag you selected).

**Comment by 1Maxnet1** [comment]
> Made the same mistake myself a couple of times ;-)
> 
> When using tags to distinguish different timetables (which is required with `nigiri`, even if you have only one dataset), you need to prefix your station IDs with those tags. The reason is that different timetable datasets sometimes use the same station ID, so a station ID would not be unique anymore without the prepended tag. In your case, the station `de:08118:7404:1:3` from your dataset would be `vvs_de:08118:7404:1:3` in MOTIS (depending on the tag you selected).

Can it do so on "on the fly" when importing the GTFS-data or am I required to do so before importing the GTFS-data?

**Comment by felixguendling** [comment]
You don't need to change the data you import at all. You can just download the ZIP file and point MOTIS to that  ZIP file `import.paths=schedule-vvs:/import/vvs.zip` (unpacking like you did it also works and saves RAM).

MOTIS automatically prepends the tag (here `vvs` + `_`) to each loaded stop ID. Therefore, you also need to prepend the same tag to the stop IDs in your query.

Those tags are also used in the URL: `https://europe.motis-project.de/#/station/delfi_de:03152:37762/1687342980`

**Comment by 1Maxnet1** [comment]
> You don't need to change the data you import at all. You can just download the ZIP file and point MOTIS to that ZIP file `import.paths=schedule-vvs:/import/vvs.zip` (unpacking like you did it also works and saves RAM).
> 
> MOTIS automatically prepends the tag (here `vvs` + `_`) to each loaded stop ID. Therefore, you also need to prepend the same tag to the stop IDs in your query.
> 
> Those tags are also used in the URL: `https://europe.motis-project.de/#/station/delfi_de:03152:37762/1687342980`

Ah yeah, got it. Thanks!

**Comment by felixguendling** [comment]
For me, your query works:

```bash
curl 'https://europe.motis-project.de/?elm=IntermodalConnectionRequest' \
  -H 'Content-Type: application/json' \
  --data-raw '{
  "destination": {
    "type": "Module",
    "target": "/intermodal"
  },
  "content_type": "IntermodalRoutingRequest",
  "content": {
    "start_type": "OntripStationStart",
    "start": {
      "station": {
        "id": "delfi_de:06411:4734:46:46",
        "name": ""
      },
      "departure_time": 1687420740
    },
    "start_modes": [],
    "destination_type": "InputStation",
    "destination": {
      "id": "nl_2510297",
      "name": ""
    },
    "destination_modes": [],
    "router": "/nigiri"
  }
}' \
  --compressed
```

I needed to change the departure time and the station IDs to match the demo data on europe.motis-project.de but otherwise, your query format seems to be correct. Are you sure, that's exactly the query you sent? And are you using v0.9.4?

**Comment by 1Maxnet1** [comment]
> And are you using v0.9.4?

Yes, made sure to use v0.9.4

> Are you sure, that's exactly the query you sent?

Yes too, also tried it when sending my request with curl, which also fails. So it might be an issue in the configuration.

I will have a look and try to check the log. When I cannot find anything further, I will report back.

**Comment by felixguendling** [comment]
> So it might be an issue in the configuration.

The request parsing does not depend on the configuration.


A station to station query like yours can also be sent directly to the `nigiri` module like this:

```
curl 'https://europe.motis-project.de' \
  -H 'Content-Type: application/json' \
  --data-raw '{
  "destination": {
    "type": "Module",
    "target": "/nigiri"
  },
  "content_type": "RoutingRequest",
  "content": {
    "start_type": "OntripStationStart",
    "start": {
      "station": {
        "id": "delfi_de:06411:4734:46:46",
        "name": ""
      },
      "departure_time": 1687520443
    },
    "additional_edges": [],
    "via": [],
    "destination": {
      "id": "nl_2510297",
      "name": ""
    }
  }
}' \
  --compressed
```

**Comment by 1Maxnet1** [comment]
So I finally could fix my motis deployment somehow (although not really knowing, what part made it actually work again). I upgraded motis to 0.10.0 and ditched my old config.ini and replaced it by a new one, that is based on the one in the 0.10.0 blog post. Thanks for the help and sorry that we could not nail down the problem further. However for me the issue got resolved, including the link to the new api documentation, so feel free to mark it as resolved :)
